### ISSUE TITLE
[Bug][kimi-k3] Long-context [PAD] (id 163839) storms + DSPARK inf/nan asserts — NaN-contaminated logits; released kimi-k3 image predates #32477; pad not stopped/filtered; allowed_special="all" makes [PAD] injectable

### ISSUE THREAD

**Issue by jetd1** [sgl-project/sglang#32968]
## Summary

Serving **Kimi-K3** on the `kimi-k3` branch with DSPARK speculative decoding, we observe two recurring production failure modes that we believe share one root cause (transient **NaN contamination of logits**, consistent with the write-side bug fixed by #32477 — which is **not in any released `kimi-k3` image**):

- **A. Mid-generation `[PAD]` token storms** — the model suddenly starts emitting the pad token (**id 163839**) as literal `[PAD]` text mid-response (in either the think channel or the content channel) and never stops until `max_tokens` or client disconnect. Worst single stream we captured: **17,256 `[PAD]` occurrences, ~2 MB of SSE, 664 s of decode**.
- **B. DSPARK CUDA assert crashes** — sporadic full-engine aborts with `probability tensor contains either inf, nan or element < 0` (TensorCompare.cu assert) raised from the DSPARK prefill/draft path, killing all TP ranks (auto-recovers after reload).

Both are bursty, correlate with **long contexts (≈90k–237k tokens)**, and are invisible in short-context testing.

## Environment

- Branch: `kimi-k3` @ `f37baeb52f` source (image lineage `lmsysorg/sglang:kimi-k3-c6ad1f26-20260729`), SGLang 0.5.16, torch 2.11.0+cu130, CUDA 13.0
- Single node, 8× Blackwell-class GPUs (SM10x), TP=8
- Launch (per the Kimi-K3 cookbook, unified recipe):
  `--tp-size 8 --context-length 1000000 --disable-custom-all-reduce --enable-symm-mem --reasoning-parser kimi_k3 --tool-call-parser kimi_k3 --speculative-algorithm DSPARK --speculative-draft-model-path <RadixArk/Kimi-K3-DSpark> --speculative-dspark-block-size 7 --enable-linear-replayssm-spec --mamba-full-memory-ratio 0.1 --mem-fraction-static 0.93 --enable-hierarchical-cache --hicache-ratio 6 --hicache-mem-layout page_first --hicache-io-backend kernel --hicache-write-policy write_back --trust-remote-code`
- Weights: official MXFP4 checkpoint (`moonshotai/Kimi-K3`), draft `RadixArk/Kimi-K3-DSpark`

## Symptom A in detail — `[PAD]` storms

Observed at gateway level over a 48 h window under real agentic traffic (large system prompts, 10–80 tools, 25k–240k-token contexts): **140 storm events in 11 bursts**, hours of complete quiet in between. Channel split: ~70% inside the think channel (`reasoning_content` / thinking deltas), ~30% in visible content.

Typical stream shape (chat completions, abridged):

```
data: {"choices":[{"delta":{"reasoning_content":" evaluating the"}}]}
data: {"choices":[{"delta":{"reasoning_content":" user's"}}]}
data: {"choices":[{"delta":{"reasoning_content":"[PAD]"}}]}
data: {"choices":[{"delta":{"reasoning_content":"[PAD][PAD]"}}]}
data: {"choices":[{"delta":{"reasoning_content":"[PAD][PAD][PAD][PAD]..."}}]}   # unbounded until max_tokens / disconnect
```

During a storm the engine's own decode logs show **`accept len: 1.00, accept rate: 0.00` for hundreds of consecutive DSPARK batches** — the draft model cannot predict the degenerate distribution, so throughput collapses to non-speculative while burning until `max_tokens` (with `max_tokens=128000` agent requests this is expensive).

Impact on agent harnesses is severe: the turn produces thinking + garbage and no tool call / no content, the harness re-prompts, the poisoned long context storms again → retry loop, each cycle re-prefilling ~100k+ tokens.

## Symptom B in detail — DSPARK assert

```
/pytorch/aten/src/ATen/native/cuda/TensorCompare.cu:109: _assert_async_cuda_kernel:
Assertion `probability tensor contains either `inf`, `nan` or element < 0` failed.
...
terminate called after throwing an instance of 'c10::AcceleratorError'
```

Raised under the DSPARK worker prefill path under long-context load; whole TP group dies (SIGQUIT cascade). ~1 event/day/8-GPU node at our traffic. `python/sglang/srt/speculative/dspark_components/dspark_draft.py` already anticipates exactly this ("an all-NaN row would otherwise make multinomial raise"), so NaN reaching sampling is a known-possible state.

## Why we believe these are the same bug — mechanism chain

1. **#32477 describes the write-side defect**: CUDA-graph padding rows carry undefined K/V yet target the reserved padding slot 0; page-table padding points at that page; a partially valid tile loads the page before masking → `0 × NaN` contaminates attention output. That lands NaN in logits *transiently, depending on batch composition* — matching our "mid-response, one bad step, then normal again" signature.
2. **NaN logits in sampling are UB** — `sanitize_nan_logits()`'s own docstring: NaN logits "can come back as **out-of-vocab token ids**". The guard is **default-off** (`SGLANG_SANITIZE_NAN_LOGITS=False`).
3. **`[PAD]` = id 163839 = `vocab_size − 1`** — precisely the boundary index a corrupted argmax/top-k walk lands on. (Related precedent: #32396 fixed NaN mis-selection in the EAGLE fused top-k=1 kernel; the AITER greedy path has a comparable guard for all-NaN rows returning `vocab_size`.)
4. The same NaN reaching DSPARK's `torch.multinomial` produces Symptom B. Two exits, one defect.

**Cross-validation from production:** we cherry-picked `ee678910f7` (#32477) onto `f37baeb52f` and rolled it out worker-by-worker today. During the partial-rollout window, **every further storm landed on not-yet-patched workers; patched workers have recorded zero storms so far** (early data, hours; we will update this issue with 48 h A/B numbers against the 140-events/48 h baseline).

## Two aggravating factors independent of the root cause

**(a) `[PAD]` is not a stop token and is not filtered.** K3's `generation_config.json` only sets `eos_token_id: 163586`; pad 163839 is not in stop ids, and the serving layer must run `skip_special_tokens=False` (the XTML think markers need to survive detokenization). Net effect: one degenerate sample doesn't self-terminate — it streams literal `[PAD]` text to clients unboundedly. A single mis-sampled token becomes a 128k-token incident.

**(b) `allowed_special="all"` makes special tokens injectable from user text.** The checkpoint tokenizer (`tokenization_kimi.py`) encodes with `allowed_special="all"`, so a user message *containing the literal string* `[PAD]` tokenizes to the real pad id 163839 (verified). Untrusted input can therefore seed real special tokens into the context — and because storm output that reaches conversation history is replayed on the next turn, there is also a self-contamination loop (later turns imitate the degenerate pattern).

## Suggested actions

1. **Publish a new `kimi-k3` image containing #32477** (current latest `kimi-k3-c6ad1f26-20260729` predates it; the branch HEAD already has it via the main merges). For long-context K3 + DSPARK this appears to be correctness-critical, not just hygiene.
2. **Treat pad as a hard stop / filter for K3 serving**: add 163839 to default stop ids for `kimi_k3` (or drop pad tokens in the detokenizer even when `skip_special_tokens=False`). Either bounds the blast radius of any future sampling corruption to one token.
3. **Consider defaulting `SGLANG_SANITIZE_NAN_LOGITS=1`** (or at least recommending it in the K3 cookbook) — with the caveat that for a *fully*-NaN row it converts the failure into a uniform-random sample; a cheap boundary-id guard in the sampling kernels (like #32396 did for EAGLE) may be the better structural fix.
4. **Reconsider `allowed_special="all"` for untrusted chat text** (encode user content with special-token strings escaped), keeping structural tokens injectable only by the chat encoder itself.

Happy to provide additional stream captures, decode-log excerpts, or run instrumented builds — and we'll report the post-#32477 A/B numbers here either way.

**Comment by irony** [comment]
Confirming this on our production K3 deployment — 8× NVIDIA B300 SXM6, single node, TP=8, `lmsysorg/sglang:kimi-k3-c6ad1f26-20260729-amd64` (same image lineage as OP, contains #32571 but predates #32477).

## Our repro: low `max_tokens` is sufficient to trigger

Unlike OP's long-context correlation, we can reproduce the PAD storm with **trivially short prompts and low `max_tokens`**:

```bash
curl -X POST http://localhost:30013/v1/chat/completions \
  -d '{"model":"moonshotai/Kimi-K3","messages":[{"role":"user","content":"Say: hello"}],"max_tokens":50,"temperature":0}'
```

→ `finish_reason: "length"`, content ~44 tokens then PAD-loop until max_tokens exhausted. With `max_tokens=500` the same request returns `finish_reason: "stop"` normally. This points to the reasoning phase hitting a block boundary mid-CoT — exactly the write-side defect #32477 addresses.

(k3 always reasons via `--reasoning-parser kimi_k3`; the reasoning tokens consume most of a low `max_tokens` budget, so the model is mid-block when `max_tokens` cuts in.)

## Production signature: hung requests + bursty PAD storms

- Two requests hung in an **unbounded PAD-loop for 40+ minutes** — engine logs show `#running-req: 2` unchanged, `gen throughput: ~93 tok/s` but effectively all PAD (full token count barely creeping, +2.5 tok/s effective).
- The PAD production never self-terminates because PAD (163839) isn't in stop ids — a single bad sample becomes a multi-hour incident confined only by `max_tokens`/client disconnect.
- Bursty pattern matches OP exactly: PAD-storms hit multiple users simultaneously (the hung requests back up the queue), then quiet for hours. We initially misdiagnosed this as a client-side issue.

## Adds weight to suggested actions 2 + 3

- **Action 2 (pad as hard stop/filter)** would have contained each of our incidents to a single token. We're planning to add 163839 to stop ids client-side as immediate mitigation.
- **Action 3 (`SGLANG_SANITIZE_NAN_LOGITS=1`)** — we'll try this too and report back if it helps.

Happy to provide engine-side decode logs or `accept_len: 1.00, accept_rate: 0.00` batch excerpts if useful.

**Comment by jetd1** [comment]
Promised follow-up, and one important caveat-turned-real.

**Update on Suggested Action 3 (`SGLANG_SANITIZE_NAN_LOGITS=1`):** we enabled it fleet-wide alongside the #32477 cherry-pick. Overnight results:

- **Zero device-side asserts / engine crashes** since enabling — that axis works as designed.
- **≥3 full-row NaN events across 3 of 4 nodes** (22:26Z / 23:07Z / 23:34Z), each logging `NaN detected in sampler: next_token_logits …` on all 8 TP ranks simultaneously, at first-token sampling after a prefill (the ones we correlated show small 64-token, cached=0 prefills). Speculative decoding is **disabled** on this fleet now — so this NaN source is on the plain prefill path, independent of DSPARK and not covered by #32477's decode-graph fix. "≥" because `_AsyncNanWarner` is a process-lifetime one-shot latch; repeat events on the same worker are silent.
- The caveat we flagged in Action 3 materialized in production: a full-row NaN event under the flag becomes a **uniform-random sample over the whole vocab** — one of our real user sessions received multilingual token salad (including a literal `<|reserved_token_163672|>`) streamed as a normal 200 response. Details and a proposal for an opt-in **per-request abort** mode (full-row NaN → abort only the affected request with a retriable 5xx + cache-hygiene so poisoned KV can't be shared via prefix cache): → https://github.com/sgl-project/sglang/issues/33187. PR to follow.

@irony — before you flip `SANITIZE_NAN_LOGITS=1` in prod, worth reading the linked issue first: it does eliminate the crashes, but it converts full-row NaN events from a loud failure into silent garbage in the visible stream (and partial-row events become entirely invisible). It also changes the signature you're tracking: with the flag on, the boundary-id `[PAD]` pattern largely disappears and full-row events surface as random salad instead — easy to misattribute to "model quality". The pad-as-stop-id / client-side filter mitigation from Action 2 remains worthwhile either way.

48 h A/B numbers on the [PAD]-storm baseline still coming as promised.

**Comment by llc-kc** [comment]
with this https://github.com/sgl-project/sglang/pull/32477 (without SGLANG_SANITIZE_NAN_LOGITS=1), we  still encounter crashes by /pytorch/aten/src/ATen/native/cuda/TensorCompare.cu:109: _assert_async_cuda_kernel: block: [0,0,0], thread: [0,0,0] Assertion `probability tensor contains either `inf`, `nan` or element < 0` failed.

**Comment by Haoxi2002** [comment]
Hi, do you have a solution for this issue? I also encountered this issue when deploying Kimi-k3 on B300.

**Comment by hernaninverso** [comment]
Same failure signature on a **different model and different silicon**, which may help
narrow the cause: **Qwen3.8-Flash-Next-NVFP4 (SGLang TP2) on 2× NVIDIA DGX Spark (GB10 /
SM121)**, unified 128 GB memory per node.

Identical assert, three times in one evening, always under real agent traffic:

```
/pytorch/aten/src/ATen/native/cuda/TensorCompare.cu:109: _assert_async_cuda_kernel:
Assertion `probability tensor contains either `inf`, `nan` or element < 0` failed.
```

**Matches the long-context correlation in the OP.** Full token counts at crash time:
115,840 and 120,384 — squarely inside the ~90k–237k band reported here. Short-context
benchmarking never reproduced it: three separate 3-run benchmark sweeps at <1k prompt
tokens were clean, and a 26k-token probe was clean too.

**Draft acceptance was NOT a common factor** — worth noting since it differs from the
`accept rate: 0.00` storms described above. Our two captured crashes had accept rates of
**0.82** and **0.43** in the decode batches immediately preceding the assert.

**What was common: a large chunked prefill immediately before the assert.** The last
seconds of one crash:

```
00:52:41 Decode batch  #full token: 120384  accept len: 2.30, accept rate: 0.43  mamba usage: 0.33
00:52:48 Prefill batch #new-token: 4096  #cached-token: 0     <- 11 chunks, ~45k tokens
   ...   (chunked_prefill_size=4096)
00:53:03 Prefill batch #new-token: 1792
00:53:04 POST /v1/chat/completions 200 OK
00:53:04 Assertion `probability tensor contains ...` failed
```

Config (the day-0 recipe from tonyd2wild/Qwen3.8-Flash-Next-NVFP4-DGX-Spark plus its
stable-70 flags):

```
--speculative-algorithm NEXTN --speculative-num-steps 3 --speculative-eagle-topk 1
--speculative-num-draft-tokens 4 --enable-linear-replayssm-spec
--speculative-attention-mode decode --disable-radix-cache --disable-prefill-cuda-graph
--cuda-graph-max-bs 8 --disable-cuda-graph-padding --sampling-backend pytorch
--quantization modelopt_fp4 --context-length 262144 --mem-fraction-static 0.80
--max-total-tokens 600000 --max-running-requests 6 --chunked-prefill-size 4096
--json-model-override-args '{"language_model_only": true}'
```

Reproduced with **and** without `--disable-radix-cache`, and with **and** without YaRN
context extension — so neither is the trigger, at least here.

Since this is a hybrid (Mamba/SSM) model, one hypothesis worth testing is an interaction
between chunked prefill and the SSM state replay used by speculative verification
(`--enable-linear-replayssm-spec`). We have not isolated it.

Enabling `SGLANG_SANITIZE_NAN_LOGITS=1` now, per @jetd1's overnight results above, and
accepting the documented trade-off (uniform-random sample instead of an engine crash).
Happy to report back with 24 h of data from this hardware if useful.

**Comment by hernaninverso** [comment]
Follow-up to my report above (Qwen3.8-Flash-Next-NVFP4, SGLang TP2, 2× DGX Spark GB10 / SM121).

**Correction to my own data point: I can now reproduce it with a short prompt.** In my earlier comment I said short-context benchmarking never reproduced it and that our crashes sat squarely in the ~90k–237k band. That band held for the *hard crashes*, but with `SGLANG_SANITIZE_NAN_LOGITS=1` enabled the failure changes shape and shows up at trivial prompt sizes too. This lines up with @irony's and @jetd1's reports rather than with the long-context correlation in the OP.

**Repro:** 5-word prompt, `max_tokens: 12`, no system prompt, through a plain `/v1/chat/completions` call:

```
prompt:   "Responde solo: OK"   (i.e. "reply only: OK")
response: 'OK складаระ并被渝 Keg治愈 cevap Hans Female当然是 nỗ'
```

The response starts correctly and then degenerates into a mix of unrelated scripts (CJK + Cyrillic + Thai + Korean + Latin) — the sanitized-NaN signature, where a full-NaN logits row becomes a uniform draw over the vocabulary. In longer agent turns the same event runs until `max_tokens` is exhausted; the worst ones we logged burned the full 32,000-token budget and ~15 minutes of wall clock before returning.

**It is intermittent, not constant.** Right after capturing the above I ran 24 consecutive probes (8 per lane, 3 lanes, varied prompt to defeat any caching) and got **0 degenerate responses**, median latency 0.2 s. So single-digit sample sizes will happily show either 1/3 or 0/24 — worth noting for anyone trying to bisect this by eye.

**Configuration, verified in the running container rather than assumed:**

- `#32477` is present — `reserved_skip_index` appears in `python/sglang/kernels/jit/csrc/elementwise/kvcache.cuh`. So this is *not* the write-side bug that PR fixes.
- The QSA resolver fix from `#36556` is applied (needed at all on SM121, see `#36558`), so decode is on the FlashInfer TRTLLM path, not the CuTe fallback.
- `SGLANG_SANITIZE_NAN_LOGITS=1`.
- Unlike @jetd1's fleet, **speculative decoding is enabled here** (`NEXTN`, steps=3, topk=1, draft=4) and we still see it — consistent with their conclusion that this NaN source is independent of the speculative path.

**One possibly useful correlation.** @jetd1 noted the events they correlated were *"small 64-token, `cached=0` prefills"*. We run with **`--disable-radix-cache`**, which forces `cached=0` on *every* prefill, on every turn — we cannot get a cached prefill even in principle. If the NaN really is on the plain prefill path at first-token sampling, a deployment with the prefix cache disabled is permanently in the worst case, which would explain why we hit this often enough to notice under ordinary agent traffic.

That gives a testable prediction for anyone who can run it: **if the rate drops when the radix cache is enabled, the trigger is the uncached prefill path.** We can't run that A/B on our side right now — this deployment previously died at startup without `--disable-radix-cache`, though that was on an image predating the `#36556` resolver fix, so the two may not be independent.

Happy to run instrumented probes on SM121 / GB10 if a specific counter or log line would help narrow it down.
